api/controllers/CheckUploadedDocumentsController.js

```diff
@@ -38,11 +51,149 @@ const CheckUploadedDocumentsController = {
             ? uploadedFile.storageName
             : uploadedFile.location;
 
+        if (!sessionData.appId) {
+            throw new Error('Missing applicaiton id');
+        }
         return {
-            application_id: sessionData.appId || 0, // TODO throw error if this value is false
+            application_id: sessionData.appId,
             uploaded_url: fileUrl,
         };
     },
+
+    _checkDocumentCountAndPaymentDetails(req, res) {
+        const { appId, eApp, payment_reference: paymentRef } = req.session;
+        const documentCount = eApp.uploadedFileData.length;
+        const totalPrice = documentCount * COST_PER_DOCUMENT;
+        const redirectUrl = req._sails.config.payment.paymentStartPageUrl;
+        const paymentParams = { appId, totalPrice, paymentRef, redirectUrl };
+        const documentCountParams = {
+            appId,
+            totalPrice,
+            documentCount,
+        };
+        CheckUploadedDocumentsController._checkDocumentCountInDB(
+            documentCountParams,
+            res
+        );
+        // Redirect to payment handled here
+        CheckUploadedDocumentsController._checkPaymentDetailsExistsInDB(
+            paymentParams,
+            res
+        );
```

Comment:
Because these calls are async I'd chain them, or at least in some way make sure that they're not running in parallel, if soemthing goes wrong in `_checkDocumentCountInDB` we still call `_checkPaymentDetailsExistsInDB`

Reply:
Yes, that's a good point. I will do that.

Reply:
Have thought of another thing to consider on this - what you want the behavious to be if `_createDocumentCountInDB` but `_createPaymentDetailsInDB` fails. Currently the doc count record would be in the database and the payment details record wouldn't be. From the looks of the code this probably doesn't matter so much, user will hopefully resubmit and the second time around the count record will be updated rather than created and all will be well in the world.

If it does matter though - there may be other areas of the app where having one record without the other will stuff things up. Or maybe just to maintain data integrity ou don't want incomplete submissions. You can wrap the sequalize calls in a transaction so if any fail you rollback all and keep the DB clean for the next run.

Reply:
Interesting, that's something I don't have a lot of experience with, Ali mentioned it to me today actually. I'll take a look at that too.

- [ ] Look into transactions to keep data clean

Reply:
Actually, now that I'm looking at the code again I think it's not that big of a deal. As in, the document count can be created or updated without the payment details and vice versa. However, both bits of information are needed before the user can pay it doesn't really matter what order they go in.

So technically the user would try again, and eventually, call if they can't submit. Something to test in the MVP I guess 🤷‍♂️